src/coreclr/jit/compiler.cpp

```diff
@@ -5360,6 +5360,23 @@ void Compiler::generatePatchpointInfo()
             }
         }
 
+        // If the this pointer can be modified, the patchpoint should record
+        // the location for the modifiable this.
+        //
+        // The Tier0 method's scratch BB will have code (see `fgAddInternal`) that
+        // ensures this location always includes an appropriate `this` (either original
+        // or modified) so we don't need to worry about the relative timing of
+        // the potential modification by Tier0 vis-a-vis the timing of any OSR transition.
+        //
+        // Note when the orginal `this` must be kept alive (for say generics context
+        // reporting) its location is conveyed by a separate field in the patchpoint info.
+        // The OSR method will not need to refer to the original Tier0 `this` in codegen.
+        //
+        if ((lclNum == 0) && !info.compIsStatic && !lvaIsOriginalThisReadOnly())
+        {
+            varNum = lvaArg0Var;
+        }
```

Comment:
Is there a risk that this ends up changing the generic context if we are fetching that through the this pointer? Will the OSR version suddenly see a new generic context?

Reply:
For example:
```csharp
class CBase<T>
{
  void Foo()
  {
    this = new CDerived<string, T>();
    // OSR triggers here
    Console.WriteLine(typeof(T).FullName);
  }
}

class CDerived<T1, T2> : CBase<T2>
{
}
```

May need some finessing...

Reply:
Yes, seems like that could happen, `getRuntimeContextTree` explicitly refers to the unmodified `this`.

Might require IL to create a suitable example.

Reply:
`[MethodImpl(MethodImplOptions.Synchronized)]` may have the same problem if that is an easier test to write.

Reply:
I haven't come up with anything yet. C# doesn't allow assignment to this (for ref class instance methods anyways).

I suppose I could look at the F# example more closely.